xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/ConversionMatrix.h

```diff
@@ -18,34 +20,85 @@ template <unsigned Order>
 class CMatrix
 {
 public:
-  CMatrix(float (&src)[Order][Order]);
-  CMatrix(float (&src)[Order-1][Order-1]);
-  CMatrix& operator=(const CMatrix& src);
-  CMatrix& operator=(const float (&src)[Order-1][Order-1]);
+  CMatrix() = default;
+  CMatrix(CMatrix<Order - 1>& other);
+  CMatrix(std::array<std::array<float, Order>, Order>& other) { m_mat = other; }
+  CMatrix(std::array<std::array<float, Order - 1>, Order - 1>& other);
   virtual ~CMatrix() = default;
 
-  float (&Get())[Order][Order];
-  CMatrix Invert();
+  virtual CMatrix operator*(const std::array<std::array<float, Order>, Order>& other);
+
   CMatrix operator*(const CMatrix& other);
   CMatrix operator*=(const CMatrix& other);
-  virtual CMatrix operator*(const float (&other)[Order][Order]);
+
+  CMatrix& operator=(const std::array<std::array<float, Order - 1>, Order - 1>& other);
+
+  bool operator==(const CMatrix<Order>& other) const
+  {
+    for (unsigned i = 0; i < Order; ++i)
+    {
+      for (unsigned j = 0; j < Order; ++j)
+      {
+        if (m_mat[i][j] == other.m_mat[i][j])
+          continue;
+
+        // some floating point comparisions should be done by checking if the difference is within a tolerance
+        if (std::abs(m_mat[i][j] - other.m_mat[i][j]) <=
+            (std::max(std::abs(other.m_mat[i][j]), std::abs(m_mat[i][j])) * 1e-2))
+          continue;
+
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  std::array<float, Order>& operator[](int index) { return m_mat[index]; }
+
+  std::array<std::array<float, Order>, Order>& Get();
+
+  CMatrix Invert();
+
+  float* ToRaw()
+  {
+    float* raw{nullptr};
+
+    for (unsigned i = 0; i < Order; i++)
+    {
+      for (unsigned j = 0; j < Order; j++)
+      {
+        raw = &m_mat[i][j];
+        raw++;
+      }
+    }
+
+    raw -= Order * Order;
+    return raw;
+  }
+
+  void Reset() { m_initialized = false; }
+  void SetInitialized() { m_initialized = true; }
+  bool IsInitialized() { return m_initialized; }
```

Comment:
Should this really be public?

Reply:
I'm not sure how else to handle it :/

Reply:
I think this concern doesn't belong in `CMatrix`. How about moving the `bool` (or two `bool`s) to `CConvertMatrix`?

Reply:
problem is that just a few lines down from here these methods are also used on `m_matPrim`

Reply:
would a `std::optional<CMatrix>` work in `CConvertMatrix`?

Reply:
we aren't `c++17` yet 😉

Reply:
haha sorry, I've been doing c++20 at work lately, I might be a bit out of touch :joy: 

The behavior you are going for is `std::optional` as far as I would interpret it, but we don't have that. So IMO best thing to do would still be to rip these methods out of this class and put two `bool`s in `CConvertMatrix`, one each for `m_mat` and `m_matPrim` and replace the method calls with the code. It wouldn't really change the length of the code and gets rid of a concern that doesn't belong in `CMatrix`. But that's only a suggestion.

Reply:
what about using `std::unique_ptr` to check/reset and get rid of the custom methods?

Reply:
that sounds like a good approximation of the behavior and we can neglect the overhead

Reply:
updated, not sure if I like passing the reference of a `unique_ptr` though.

Reply:
it's not by itself any more or less evil to return the reference of something stored as `unique_ptr` than something stored as anything else (like a member variable). if you get something as a reference, you should assume that you do not have ownership.
Either way, `auto mat = GenMat();` will do a copy so you wouldn't need the reference in the first place :D

Reply:
damn you're right, I'll fix that.

Reply:
updated